### DISCUSSION TITLE
Migrator up with bun is not working

### DISCUSSION THREAD

**Discussion by co-sic** [mikro-orm/mikro-orm#4784] [Q&A]
I guess this is a bun issue and not mikro-orm, but maybe somebody has an idea how to solve this. I'm using bun v1.0.4 with mikro-orm v5.8.6 and postgres driver. Everything is working, expect of the migrator. I get the following error:

```
[query] select table_name, table_schema as schema_name, (select pg_catalog.obj_description(c.oid) from pg_catalog.pg_class c where c.oid = (select ('"' || table_schema || '"."' || table_name || '"')::regclass::oid) and c.relname = table_name) as table_comment from information_schema.tables where "table_schema" not like 'pg_%' and "table_schema" not like 'crdb_%' and "table_schema" not like '_timescaledb_%' and "table_schema" not in ('information_schema', 'tiger', 'topology') and table_name != 'geometry_columns' and table_name != 'spatial_ref_sys' and table_type != 'VIEW' order by table_name [took 9 ms, 1 result]
[query] begin
[query] select * from "public"."mikro_orm_migrations" order by "id" asc [took 2 ms, 0 result]
Processing 'Migration20230304144659'
[query] rollback
MigrationError: Migration Migration20230304144659 (up) failed: Original error: undefined is not a constructor (evaluating 'new (Object.values(migration))[0](this.driver, this.config)')
 info: {
  "direction": "up",
  "name": "Migration20230304144659",
  "path": "[...]/migrations/Migration20230304144659.ts",
  "context": {}
}
```

I found out the following two things:
### First
node_modules look like this after a clean bun install:
```
@mikro-orm
-core
-knex
-migrations
-postgresql
-reflection
```
If i change it to add a node_modules folder under migrations with core like this, then the migration will work:
```
@mikro-orm
-core
-knex
-migrations
--node_modules
---@mikro-orm
----core
-postgresql
-reflection
```
i found this out since currently bun install does not handle upgrading the version correctly. If i change the version of mikro-orm to some other version and run `bun install` it will install the versions correctly on the root @mikro-orm/*** level but will install some other versions of the peer dependencies inside the node_modules folder of the mikro-orm packages. But this is not correct and also cause typescript errors, so currently i just delete the bun lock file and the node modules folder and run `bun install`, then the structure is flat again without additional versions being installed.

### Second
If i go inside the Migrator class and log the migrations object from the dynamic Import i get:
```
Module {
  __esModule: true,
  Migration20230305154009: [class Migration20230305154009]
}
```
and if i log `Object.values(migration)` i get `[]`.
If i change `const MigrationClass = Object.values(migration)[0];` to `const MigrationClass = migration.Migration20230304144659` then it will also work for this specific migration file of course.

**Comment by B4nan** [comment]
Can you share your setup? I was actually trying out bun few days ago and it didn't really work for me, but I guess I only tried (both) sqlite drivers.

The second problem seems to be worth reporting to bun, I believe the first one is a known issue, saw something like that recently.

**Reply by co-sic** [reply]
I can try to make a minimal version of the setup tomorrow and share it :+1:  
I converted one of our existing mikro-service to bun and it is actually running in produciton already (it's not a critical service). I just noticed later that the migration is not working, since it was working sometimes with the first issue i pointed out, so i thought everything was fine. So until i need a new migration i can keep it running on bun, but then i need to find a solution or rollback to node again.

**Reply by co-sic** [reply]
@B4nan i changed this repo here i used for a previous issue to use bun. You can check the Readme for the workarounds i needed to do: https://github.com/co-sic/node-mikro-orm-playground

**Reply by WDaan** [reply]
Thanks @co-sic that's the only way i've gotten it to work.

**Reply by emulienfou** [reply]
Thanks @co-sic for this temporary workarounds, had the same issue with:
- Node v18.17.1
- @mikro-orm/core v6.0.0-dev.193

I was able to fix it by changing only this 2 lines of code:
```ts
resolve(params) {
        // Add this line
        const fileName = params.path.split('/').pop().split('.')[0];
        const createMigrationHandler = async (method) => {
        const migration = await core_1.Utils.dynamicImport(params.path);
        // Comment this line:
        // const MigrationClass = Object.values(migration)[0];
       // Adding this line instead:
          const MigrationClass = Object.values(migration)[0][fileName];
            const instance = new MigrationClass(this.driver, this.config);
            await this.runner.run(instance, method);
        };
        return {
            name: this.storage.getMigrationName(params.name),
            up: () => createMigrationHandler('up'),
            down: () => createMigrationHandler('down'),
        };
    }
```

**Comment by co-sic** [comment]
https://github.com/mikro-orm/mikro-orm/discussions/5349 this one has a better workaround for now
